### PR TITLE
Spring cleaning in summer 2024

### PR DESCRIPTION
**Pull request by djeremiah**
### Description of your changes
- update build submodule from https://github.com/upbound/build to https://github.com/crossplane/build
- update various dependencies 
    - go 1.21
    - crossplane-runtime 1.16
    - crossplane-tools 20240522
    - go-cmp 0.6.0
    - k8s.io/{api,apimachinery,client-go} 0.29.3
    - controller-runtime 0.17.1
    - controller-tools 0.14.0
- enable high-level MR metrics
- update the linter config to match c/c and fix lint errors

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Local testing using the existing example nop resource

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -103,35 +107,21 @@ linters-settings:
...
     - govet
     - gocyclo
     - gocritic
+    - interfacer
     - goconst
     - gci
     - gofmt  # We enable this as well as goimports for its simplify mode.
     - prealloc
-    - revive
+    - golint
```
Comment (negz) [inline .golangci.yml]: Golint was deprecated and replaced by revive. So I think this change will effectively be disabling a linter.

What's the goal behind the linter changes here? If you want to get this repo up-to-date with the latest and greatest linter config, I'd recommend copying the one from c/c.

File: Makefile [hunk 2]
```diff
@@ -28,8 +28,8 @@ NPROCS ?= 1
 # to half the number of CPU cores.
 GO_TEST_PARALLEL := $(shell echo $$(( $(NPROCS) / 2 )))
 
-GO_REQUIRED_VERSION ?= 1.19
-GOLANGCILINT_VERSION ?= 1.53.3
+GO_REQUIRED_VERSION ?= 1.21
```
Comment (negz) [inline Makefile]: I don't think anything reads this variable anymore. We let go.mod handle it. https://github.com/search?q=repo%3Acrossplane%2Fbuild+GO_REQUIRED_VERSION&type=code

**Review by negz — COMMENTED** [review]
Two nits, but this looks good overall. Thanks for doing this!

File: Makefile (hunk 2 above)
Comment (djeremiah) [reply Makefile]: Roger. I'll delete it.

File: .golangci.yml (hunk 1 above)
Comment (djeremiah) [reply .golangci.yml]: I copied most of the build metadata from provider-kubernetes to match what I assumed was a more up-to-date provider. I'll target c/c instead.

**Review by negz — APPROVED** [review]
Looks good to me, thank you!

**Comment by negz** [comment]
@djeremiah Looks like this is failing because I'd removed the `make vendor` target from the build submodule. It was re-added in https://github.com/crossplane/build/pull/6, so I think updating the submodule one more time should make this build.

Edit: Or even better, update e.g. https://github.com/crossplane-contrib/provider-nop/blob/main/.github/workflows/ci.yaml#L73 to use the new `module.download` and `modules.check` targets.

**Comment by djeremiah** [comment]
@negz Thanks! I've gone ahead and done both. Updated the build submodule to latest and also updated all of the .github/ folder to match the latest in crossplane/provider-template

Can you approve the workflow to confirm it's working?

**Comment by negz** [comment]
@djeremiah Looks like lint is failing because the linter config from c/c is for a newer version of golangci-lint. Crossplane is using v1.59 - https://github.com/crossplane/crossplane/blob/691dcc/Earthfile#L200. I think v1.59 requires Go 1.22 - hopefully bumping both of those versions will make this pass CI.

**Comment by djeremiah** [comment]
@negz I was confused, because I ran the linter locally as part of the updates I did, but it turns out the ci.yml doesn't use `make lint`, it runs with its own configuration. This means the version needs to be specified in two places: Makefile and ci.yml. I have matched them up, and added a note to the Makefile to remind future maintainers to keep them in sync.

**Pull request merged**